### PR TITLE
Switch from lvm2app.h to calling the LVM2 CLI utils

### PR DESCRIPTION
**Pull request by gpaul**
Yay!

This makes it very easy to add support for custom command-line flags to be specified and passed to vgcreate or lvcreate by the user. It also makes the plugin as portable as any other Go project (i.e., YMMV.)

One interesting consequence of having removed our lvm2app.h integration is that we no longer lock around all operations - I don't know if that's a problem. We don't store any state in LVM2 so there shouldn't be any races - I trust that LVM2's command-line utilities are re-entrant, too. 

It might be safer to just wrap a mutex around all HTTP requests, but that's pretty grim.

This is a big PR.

https://jira.mesosphere.com/browse/DCOS_OSS-2276

### CODE REVIEW

**Comment by gpaul** [comment]
@jieyu @jdef should we merge this?

**Comment by jdef** [comment]
Thanks for this PR, I haven't had time to review it yet - trying to land
the volume-delete stuff. This is in my backlog!

File: pkg/lvm/lvm2app.go [hunk 1]
```diff
@@ -892,23 +574,112 @@ func ListVolumeGroupNames() ([]string, error) {
...
+	return nil, ErrPhysicalVolumeNotFound
 }
 
 // Extent sizing for linear logical volumes:
 // https://github.com/Jajcus/lvm2/blob/266d6564d7a72fcff5b25367b7a95424ccf8089e/lib/metadata/metadata.c#L983
+
+func run(cmd string, v interface{}, extraArgs ...string) error {
+	var args []string
+	if v != nil {
+		args = append(args, "--reportformat=json")
```
Comment (jdef) [inline pkg/lvm/lvm2app.go]: what's the min version requirement for LVM in order to get JSON report format? for example:
```sh
$ uname -a
Linux node-1 4.4.0-116-generic #140-Ubuntu SMP Mon Feb 12 21:23:04 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

$ dpkg -l|grep lvm
...
ii  lvm2                                2.02.133-1ubuntu10                         amd64        Linux Logical Volume Manager

$ sudo lvs --reportformat=json
lvs: unrecognized option '--reportformat=json'
  Error during parsing of command line.
```

we might want to include a min version check for LVM tooling at plugin startup and fail immediately if it's too old

File: pkg/lvm/lvm2app.go [hunk 2]
```diff
@@ -892,23 +574,112 @@ func ListVolumeGroupNames() ([]string, error) {
...
+	log.Printf("stderr: " + errstr)
+	if v != nil {
+		if err := json.Unmarshal(stdoutbuf, v); err != nil {
+			return fmt.Errorf("%v: [%v]", err, string(stdoutbuf))
+		}
+	}
+	return nil
+}
+
+func ignoreWarnings(str string) string {
```
Comment (jdef) [inline pkg/lvm/lvm2app.go]: nit: this isn't very efficient. although it does the job, we may want to refactor this at some point

File: pkg/lvm/lvm2app_test.go [hunk 3]
```diff
@@ -166,44 +109,20 @@ func TestLookupPhysicalVolumeNonExistent(t *testing.T) {
...
+	if err := ValidateTag(strings.Repeat("a", 1024)); err != nil {
 		t.Fatalf("Expected tag to pass validation")
 	}
-	if err := handle.ValidateTag("some:tag"); err != ErrTagHasInvalidChars {
+	if err := ValidateTag("some:tag"); err != ErrTagHasInvalidChars {
 		t.Fatalf("Expected tag to fail validation")
 	}
 }
 
-func TestValidateVolumeGroupName(t *testing.T) {
```
Comment (jdef) [inline pkg/lvm/lvm2app_test.go]: why did we delete the test for vg name validation?

File: pkg/lvm/lvm2app_test.go [hunk 4]
```diff
@@ -365,8 +257,8 @@ func TestCreateVolumeGroupInvalidName(t *testing.T) {
 		t.Fatal("Expected no volume group in response")
 	}
 	// Perform a known good operation to ensure that the error was
-	// cleared from the handle.
-	if _, err := handle.ListPhysicalVolumes(); err != nil {
+	// cleared from the
```
Comment (jdef) [inline pkg/lvm/lvm2app_test.go]: do we still need this? maybe it applied when we had a lib handle that cached errors, but now that we're using CLI...

**Comment by jdef** [comment]
> One interesting consequence of having removed our lvm2app.h integration is that we no longer lock around all operations - I don't know if that's a problem. We don't store any state in LVM2 so there shouldn't be any races - I trust that LVM2's command-line utilities are re-entrant, too.

For simplicity I'm in favor of leaving the locking up to LVM. `man lvm.conf` yields:

> global - Global settings
> ...
> locking_type - What type of locking to use. 1 is the default, which use flocks on files in locking_dir (see below) to avoid conflicting LVM2 commands running concurrently on a single machine. 0 disables locking and risks corrupting your metadata. If set to 2, the tools will load the external locking_library (see below). If the tools were configured --with-cluster=internal (the default) then 3 means to use built-in cluster-wide locking. Type 4 enforces read-only metadata and forbids any operations that might want to modify Volume Group metadata. All changes to logical volumes and their states are communicated using locks.

**Comment by gpaul** [comment]
Thanks @jdef I agree regarding leaving locking up to LVM and mentioning the fact.

File: cmd/csilvm/csilvm.go [hunk 5]
```diff
@@ -26,7 +27,8 @@ func (f *stringsFlag) String() string {
 }
 
 func (f *stringsFlag) Set(tag string) error {
-	*f = append(*f, tag)
+	*f = []string{tag}
```
[the remark this answers is not in the document]
Comment (gpaul) [reply cmd/csilvm/csilvm.go]: Woops - that snuck in from a debugging session. I'll revert.

File: cmd/csilvm/csilvm.go [hunk 6]
```diff
@@ -46,6 +48,7 @@ func main() {
 	logprefix := fmt.Sprintf("[%s]", *vgnameF)
 	logflags := log.LstdFlags | log.Lshortfile
 	csilvm.SetLogger(log.New(os.Stderr, logprefix, logflags))
+	lvm.SetLogger(log.New(os.Stderr, logprefix, logflags))
```
[the remark this answers is not in the document]
Comment (gpaul) [reply cmd/csilvm/csilvm.go]: Good idea. Will do.

File: pkg/lvm/lvm2app_test.go (hunk 4 above)
Comment (gpaul) [reply pkg/lvm/lvm2app_test.go]: Good spot - we don't need this anymore.

File: pkg/lvm/lvm2app.go (hunk 2 above)
Comment (gpaul) [reply pkg/lvm/lvm2app.go]: My thinking w.r.t. performance is usually: compared to the overhead of serving an HTTP request or executing a cli command string manipulation is typically free.

File: pkg/lvm/lvm2app.go (hunk 1 above)
Comment (gpaul) [reply pkg/lvm/lvm2app.go]: That's a really good point. I never considered that the CLI interface we rely on may have changed in living memory.

File: pkg/lvm/lvm2app_test.go (hunk 3 above)
Comment (gpaul) [reply pkg/lvm/lvm2app_test.go]: That function tested an internal function that was removed. There is a separate test that checks the 'external' API for when a volume group name is bad called `TestCreateVolumeGroupInvalidName`.

**Comment by gpaul** [comment]
Remaining AI's: 
1. Check lvm2 version or figure out some other strategy for testing different versions.
2. Update README to reflect the changes we've made and to mention the fact that we rely on lvm2's locking.

**Comment by gpaul** [comment]
Also, rename `lvm2app.go` to `lvm.go` now that we don't interface with lvm2app.h anymore.

**Comment by gpaul** [comment]
OK - one left: 
> Check lvm2 version or figure out some other strategy for testing different versions.

I'm tempted to say: 
> For maximum confidence you should run the unit tests in an environment that matches your production > environment. The plugin has been tested on Centos 7.3 and is expected to work with lvm2 >= x.y.z.

The only way to really know if it works for some version of lvm2 is to run the tests for that version. Checking that support for json output has been added in some version does not convey full confidence that everything else will work in that version, nor that it won't break in some intervening version.

@jdef WDYT?

**Comment by gpaul** [comment]
@jdef thanks again for the review. PTAL

File: Dockerfile [hunk 7]
```diff
@@ -1,6 +1,6 @@
-FROM centos:7
+FROM centos:7.3.1611
```
Comment (jdef) [inline Dockerfile]: maybe a comment here along the lines of "Update the README when this base image and/or the version of lvm2 (below) is updated"?

**Pull request merged**